dbaccess/block.go

```diff
@@ -2,156 +2,144 @@ package dbaccess
-	"github.com/kaspanet/kasparov/httpserverutils"
 )
 
 // BlockByHash retrieves a block from the database according to its hash
 // If preloadedFields was provided - preloads the requested fields
-func BlockByHash(ctx Context, blockHash string, preloadedFields ...dbmodels.FieldName) (*dbmodels.Block, error) {
-	db, err := ctx.db()
+func BlockByHash(ctx database.Context, blockHash string, preloadedFields ...dbmodels.FieldName) (*dbmodels.Block, error) {
+	db, err := ctx.DB()
 	if err != nil {
 		return nil, err
 	}
 
-	query := db.Where(&dbmodels.Block{BlockHash: blockHash})
-	query = preloadFields(query, preloadedFields)
-
 	block := &dbmodels.Block{}
-	dbResult := query.First(block)
-
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.IsDBRecordNotFoundError(dbErrors) {
+	query := db.Model(block).
+		Where("block.block_hash = ?", blockHash)
+	query = preloadFields(query, preloadedFields)
+	err = query.First()
+	if err == pg.ErrNoRows {
 		return nil, nil
 	}
-	if httpserverutils.HasDBError(dbErrors) {
-		return nil, httpserverutils.NewErrorFromDBErrors("some errors were encountered when loading block from database:",
-			dbResult.GetErrors())
+	if err != nil {
+		return nil, err
 	}
 
 	return block, nil
 }
 
 // BlocksByHashes retreives a list of blocks with the corresponding `hashes`
-func BlocksByHashes(ctx Context, hashes []string, preloadedFields ...dbmodels.FieldName) ([]*dbmodels.Block, error) {
-	db, err := ctx.db()
+func BlocksByHashes(ctx database.Context, hashes []string, preloadedFields ...dbmodels.FieldName) ([]*dbmodels.Block, error) {
+	if len(hashes) == 0 {
+		return nil, nil
+	}
+
+	db, err := ctx.DB()
 	if err != nil {
 		return nil, err
 	}
 
-	query := db.Where("block_hash in (?)", hashes)
-	query = preloadFields(query, preloadedFields)
-
 	blocks := []*dbmodels.Block{}
-	dbResult := query.Find(&blocks)
-
-	dbErrors := dbResult.GetErrors()
-	if httpserverutils.HasDBError(dbErrors) {
-		return nil, httpserverutils.NewErrorFromDBErrors("some errors were encountered when loading blocks from database:",
-			dbResult.GetErrors())
+	query := db.Model(&blocks).Where("block.block_hash in (?)", pg.In(hashes))
+	query = preloadFields(query, preloadedFields)
+	err = query.Select()
+	if err != nil {
+		return nil, err
 	}
 
 	return blocks, nil
 }
 
 // Blocks retrieves from the database up to `limit` blocks in the requested `order`, skipping the first `skip` blocks
 // If preloadedFields was provided - preloads the requested fields
-func Blocks(ctx Context, order Order, skip uint64, limit uint64, preloadedFields ...dbmodels.FieldName) ([]*dbmodels.Block, error) {
-	db, err := ctx.db()
+func Blocks(ctx database.Context, order Order, skip uint64, limit uint64, preloadedFields ...dbmodels.FieldName) ([]*dbmodels.Block, error) {
+	db, err := ctx.DB()
 	if err != nil {
 		return nil, err
 	}
 
-	query := db.
-		Offset(skip).
-		Limit(limit)
+	var blocks []*dbmodels.Block
+	query := db.Model(&blocks).
+		Offset(int(skip)).
```

Comment:
This (very theoretically) can be a problem.
In theory we allow 2^64 blocks in a dag. So in theory 1 day int might not be enough.

It's probably not a problem, but let's give it a short discussion in morning meeting

Reply:
However, please see what happens if you pass negative values here(which will happen if user supplies numbers above 2^32), I guess nothing terrible, but please make sure it is.

Reply:
It seems like you missed/forgot about this comment

Reply:
In that case we return:
```
{"errorCode":422,"errorMessage":"Couldn't parse the 'skip' query parameter"}
```